File: lib_network_retrofit_rxjava/src/main/java/com/network/retrofit_rxjava/RetrofitClient.java
package com.network.retrofit_rxjava;

import com.network.retrofit_rxjava.constants.NetworkConstant;
import com.network.retrofit_rxjava.interceptor.CacheInterceptor;
import com.network.retrofit_rxjava.interceptor.HeaderInterceptor;
import com.network.retrofit_rxjava.interceptor.LoadParamsInterceptor;
import com.network.retrofit_rxjava.interceptor.LoggingInterceptor;

import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava3.RxJava3CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

public class RetrofitClient {

    private static RetrofitClient mRetrofitClient;
    private Retrofit mRetrofit;

    private RetrofitClient(){
        //设置Http缓存
//        Cache cache = new Cache(new File(NetworkApp.getInstance().getCacheDir(), "HttpCache"),
//                1024 * 1024 * 10);
        LoggingInterceptor log = new LoggingInterceptor();
//        HttpLoggingInterceptor httpLog = new HttpLoggingInterceptor();
//        httpLog.setLevel(HttpLoggingInterceptor.Level.BODY);
        OkHttpClient okHttpClient = new OkHttpClient.Builder()
//                .cache(cache)
                .addInterceptor(log)
//                .addNetworkInterceptor(httpLog)
                .addInterceptor(new HeaderInterceptor())
                .addNetworkInterceptor(new LoadParamsInterceptor())
                .addNetworkInterceptor(new CacheInterceptor())
                .retryOnConnectionFailure(true)
                .connectTimeout(30, TimeUnit.SECONDS)
                .writeTimeout(20, TimeUnit.SECONDS)
                .readTimeout(20, TimeUnit.SECONDS)
                .build();


        mRetrofit = new Retrofit.Builder()
                .baseUrl(NetworkConstant.BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava3CallAdapterFactory.create())
                .client(okHttpClient)
                .build();
    }

    public static RetrofitClient getInstance(){
        if (mRetrofitClient == null){
            mRetrofitClient = new RetrofitClient();
        }
        return mRetrofitClient;
    }

    public <T> T create(Class<T> service){
        return mRetrofit.create(service);
    }
}

File: lib_network_retrofit_coroutine/src/main/java/com/network/retrofit_coroutine/RetrofitClient.kt
package com.network.retrofit_coroutine

import com.jakewharton.retrofit2.adapter.kotlin.coroutines.CoroutineCallAdapterFactory
import com.network.retrofit_coroutine.constants.ApiConstant
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

/**
 * 创建单例的方法：
 * 1、object 类
 * 2、class类中定义伴生对象
 * 3、枚举
 * 4、懒加载 by lazy 实例化对象(内部实现 synchronize加锁，volatile关键字)
 */
class RetrofitClient private constructor() {
    private val retrofit: Retrofit = Retrofit.Builder()
        .baseUrl(ApiConstant.BASE_URL)
        .addConverterFactory(GsonConverterFactory.create())
        //添加对 Deferred 的支持
        .addCallAdapterFactory(CoroutineCallAdapterFactory())
        .client(
            OkHttpClient.Builder()
                .addInterceptor(LoggingInterceptor())
                .build()
        )
        .build()

    companion object {
        @Volatile
        private var retrofitClient: RetrofitClient? = null
        fun getInstance(): RetrofitClient {
            return retrofitClient ?: synchronized(this) {
                /*retrofitClient ?: RetrofitClient().run {
                    retrofitClient = this
                    return retrofitClient!!
                }*/
                //run with apply 函数块内使用this,also let 函数块内使用it
                //apply also 返回值是该对象本身 run with let 返回值是最后一行的结果
                /*retrofitClient?: with(RetrofitClient){
                    retrofitClient = this//with不是扩展函数，this指代不明，无法使用该关键字
                    return retrofitClient!!
                }*/
//                retrofitClient?:RetrofitClient().apply { retrofitClient = this }
                retrofitClient ?: RetrofitClient().also { retrofitClient = it }
                /*retrofitClient?:RetrofitClient().let {
                    retrofitClient = it
                    retrofitClient!!
                }*/
            }
        }
    }

    fun <T> create(service: Class<T>): T {
        return retrofit.create(service)
    }
}